tests/mve/gtest_scene.cc

```diff
@@ -0,0 +1,715 @@
+make_bundle (const std::size_t camera_count)
+{
+    mve::Bundle::Ptr bundle = mve::Bundle::create();
+    bundle->get_cameras().reserve(camera_count);
+    for (size_t i = 1; i <= camera_count; ++i)
+    {
+        mve::CameraInfo camera;
+        camera.flen = (i % 3 == 1) ? 0 : 1.0f + 2.0f / i;
+        camera.trans[0] = i - 10.0f;
+        camera.trans[1] = 1.0f / i;
+        camera.trans[2] = 10.0f - i;
+        camera.paspect = 0.5f + 1.0f / i;
+        bundle->get_cameras().push_back(std::move(camera));
+    }
+
+    return bundle;
+}
+
+bool
+views_match (mve::Scene::ViewList& lhs, mve::Scene::ViewList& rhs)
+{
+    if (lhs.size() != rhs.size())
+        return false;
+
+    for (mve::View::Ptr& left_view : lhs)
+    {
+        mve::Scene::ViewList::iterator found = std::find_if(
+            rhs.begin(), rhs.end(), [&left_view](mve::View::Ptr& right_view) {
+                return left_view->get_id() == right_view->get_id();
+            });
+        if (found == rhs.end() || left_view->get_name() != (*found)->get_name())
+            return false;
+    }
+
+    return true;
+}
+
+bool
+bundle_cameras_match (mve::Bundle::Ptr lhs, mve::Bundle::ConstPtr rhs)
+{
+    constexpr auto is_invallid = [](const mve::CameraInfo& camera)
+    {
+        return camera.flen == 0.0f;
+    };
+
+    constexpr auto match = [](float l, float r)
+    {
+        constexpr float epsilon = 1e-3f;
+        return (r == 0.0f) ? l < epsilon : (std::abs((l / r) - 1.0f) < epsilon);
+    };
+
+    const auto match_n =
+        [&match](const float l[], const float r[], std::size_t count)
+    {
+        for (std::size_t i = 0; i < count; ++i)
+            if (!match(l[i], r[i]))
+                return false;
+        return true;
+    };
+
+    if (lhs->get_cameras().size() != rhs->get_cameras().size()
+        || lhs->get_features().size() != rhs->get_features().size())
+        return false;
+
+    for (std::size_t i = 0; i < lhs->get_cameras().size(); ++i)
+    {
+        const mve::CameraInfo& left_cam = lhs->get_cameras()[i];
+        const mve::CameraInfo& right_cam = rhs->get_cameras()[i];
+
+        if (is_invallid(left_cam))
+            return is_invallid(right_cam);
+
+        if (!match(left_cam.flen, right_cam.flen)
+            || !match_n(left_cam.dist, right_cam.dist, 2)
+            || !match_n(left_cam.trans, right_cam.trans, 3)
+            || !match_n(left_cam.rot, right_cam.rot, 9))
+            return false;
+    }
+
+    return true;
+}
+
+OnScopeExit::~OnScopeExit ()
+{
+    bool exception_happened = false;
+    for (std::string path : this->paths)
+    {
+        try
+        {
+            this->unlink_recursive(path);
+        } catch (...)
+        {
+            exception_happened = true;
+            std::cout <<  "Exception during cleanup!" << std::endl;
+        }
+    }
+    if (exception_happened)
+        throw std::runtime_error("Exception during file cleanup.");
+}
+
+void
+OnScopeExit::unlink (const std::string& str)
```

Comment:
In MVE, we write `var const&` instead of `const var&`. If you wonder why, it's because `const` ALWAYS applies to the left. Consider `var const*` and `var* const` (either var is const, or the pointer is const). Consider `void my_member (void) const` (i.e. const applies to the function on its left). There is only one exception where const applies to the right (when there is no left), and there is really no need to ever use this exception.